src/drivers/imu/bosch/bmi270/BMI270.cpp

```diff
@@ -0,0 +1,918 @@
+				break;
+
+			case 1: // init ok
+				if ((status >> 5 & 1) == 1) {
+					PX4_DEBUG("BMI270: axes_remap_error");
+
+				} else if ((status >> 6 & 1) == 1) {
+					PX4_DEBUG("BMI270: odr_50hz_error");
+				}
+
+				break;
+			}
+
+		} else if ((err >> 6 & 1) == 1) {
+			PX4_DEBUG("BMI270: fifo_err");
+
+		} else if ((err >> 7 & 1) == 1) {
+			PX4_DEBUG("BMI270: aux_err");
+
+		} else {
+			PX4_DEBUG("BMI270: internal error detected %d", err >> 1 & 0xF);
+		}
+	}
+
+#endif
+
+}
+
+void BMI270::exit_and_cleanup()
+{
+	DataReadyInterruptDisable();
+	I2CSPIDriverBase::exit_and_cleanup();
+}
+
+void BMI270::print_status()
+{
+	I2CSPIDriverBase::print_status();
+
+	PX4_INFO("FIFO empty interval: %d us (%.1f Hz)", _fifo_empty_interval_us, 1e6 / _fifo_empty_interval_us);
+
+	perf_print_counter(_bad_register_perf);
+	perf_print_counter(_bad_transfer_perf);
+	perf_print_counter(_fifo_empty_perf);
+	perf_print_counter(_fifo_overflow_perf);
+	perf_print_counter(_fifo_reset_perf);
+	perf_print_counter(_drdy_missed_perf);
+}
+
+int BMI270::probe()
+{
+	// Interface selection (SPI only): Read an arbitrary register of the device, discard the read response
+	RegisterRead(Register::CHIP_ID);
+
+	const uint8_t CHIP_ID = RegisterRead(Register::CHIP_ID);
+
+	if (CHIP_ID != chip_id) {
+		DEVICE_DEBUG("unexpected CHIP_ID 0x%02x", CHIP_ID);
+		return PX4_ERROR;
+	}
+
+	return PX4_OK;
+}
+
+void BMI270::RunImpl()
+{
+	const hrt_abstime now = hrt_absolute_time();
+
+	switch (_state) {
+	case STATE::RESET:
+
+		// 0xB6 is written to the CMD register for a soft reset
+		RegisterWrite(Register::CMD, 0xB6);
+		_reset_timestamp = now;
+		_failure_count = 0;
+		_state = STATE::WAIT_FOR_RESET;
+		ScheduleDelayed(1_ms); // Following a delay of 1 ms, all configuration settings are overwritten with their reset value.
+		break;
+
+	case STATE::WAIT_FOR_RESET:
+
+		/*
+		Hardware initialization steps according to datasheet:
+		1. Disable PWR_CONF.adv_power_save and wait for 450us
+		2. Write 0x00 to INIT_CTRL
+		3. Burst write initialization file to INIT_DATA
+		4. Write 0x01 to INIT_CTRL
+		5. Wait 150ms and read register INTERNAL_STATUS for value 0b001
+		6. If step 5 passed, enter configure state
+
+
+		*/
+		if ((RegisterRead(Register::CHIP_ID) == chip_id)) {
+			PX4_DEBUG("Read from CHIP_ID register and the IDs match");
+
+			// 1. Disable PWR_CONF.adv_power_save and wait for 450us
+
+			RegisterWrite(Register::PWR_CONF, 0x00);
+			px4_udelay(450);
+
+			// 2. Write 0x00 to INIT_CTRL
+
+			RegisterWrite(Register::CONFIG1, 0x00);
+			// give it the maximum FIFO config file
+			PX4_DEBUG("attempting to upload initialization file onto BMI270");
+
+			//  3. Burst write initialization file to INIT_DATA
+
+			int res = transfer(maximum_fifo_config_file, nullptr, sizeof(maximum_fifo_config_file));
+
+			if (res == PX4_OK) {
+				RegisterWrite(Register::CONFIG1, 1);
+				PX4_DEBUG("Successfully uploaded initialization file onto BMI270");
+				px4_mdelay(150);
```

Comment:
In general we try to avoid having any sleeps in these drivers because they'll completely block the bus (which is often shared with other sensors).

An mdelay(150) on NuttX is particularly bad because it will just spin for 150 milliseconds and starve everything lower priority.